### PR TITLE
BLD: updates to the Meson build

### PR DESCRIPTION
**Pull request by rgommers**
Mostly synced from SciPy, and also addresses some warnings that show up with Meson 1.1.0

### CODE REVIEW

**Comment by mattip** [comment]
The CI failure is in the meson build on macos_arm64.

**Comment by rgommers** [comment]
Hmm, I tested that locally, but in a conda env. Shouldn't be hard to figure out. The error was:
```
Found CMake: /opt/homebrew/bin/cmake (3.24.1)
WARNING: CMake Toolchain: Failed to determine CMake compilers state
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)

numpy/meson.build:53:9: ERROR: Dependency "OpenBLAS" not found, tried pkgconfig, framework and cmake
```
also something wrong with CMake in Homebrew it looks like.

**Comment by rgommers** [comment]
It's good that `macos_arm64` failed, because it reminded me about dealing with `lapack_lite`. It turns out that in this job, OpenBLAS was installed but not found. From another [passing CI log from today](https://cirrus-ci.com/task/6731680526893056):
```
Found CMake: /opt/homebrew/bin/cmake (3.24.1)
WARNING: CMake Toolchain: Failed to determine CMake compilers state
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
```
I _think_ we want to keep this unchanged, meaning that if no BLAS/LAPACK library was found, we use the fallback library and continue. We should at least emit a clear warning though, because for many users this continues to be a performance footgun.

**Comment by rgommers** [comment]
Not really, this is pretty routine by now, the problem should be that the `.pc` file of the OpenBLAS used there is not on pkg-config's search path. 

I am bandwidth-constrained though and may not touch this for several days, so if you want to fix it, I won't complain:)

**Comment by charris** [comment]
close/reopen

File: meson.build [hunk 1]
```diff
@@ -34,10 +35,11 @@ elif cc.get_id() == 'msvc'
           'when building with MSVC')
   endif
 endif
+if not cy.version().version_compare('>=0.29.34')
+  error('SciPy requires Cython >= 0.29.34')
```
Comment (mattip) [inline meson.build:39]: Is there a "project_name" macro we could use here instead?

```suggestion
  error('NumPy requires Cython >= 0.29.34')
```

Comment (rgommers) [reply meson.build:39]: Good catch, thanks. I'll see if the current CI run passes, then will push this with a few more tweaks.

> Is there a "project_name" macro we could use here instead?

Not that I know of - but seems more complexity than it's worth anyway, and makes it harder to grep for the error string, so I'd prefer not to do that.

**Comment by rgommers** [comment]
It looks like we should start moving some more CI jobs over to Meson soon. Also, the automatic wheel build trigger is a massive waste of jobs.

This looks better now, I'll keep it like this and will add a comment that OpenBLAS is not used:
```
Library m found: YES
Found CMake: /opt/homebrew/bin/cmake (3.24.1)
WARNING: CMake Toolchain: Failed to determine CMake compilers state
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency openblas found: NO (tried pkgconfig, framework and cmake)
Program _build_utils/process_src_template.py found: YES (/Users/admin/numpy-dev/bin/python3.10 /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/numpy/_build_utils/process_src_template.py)
Program _build_utils/tempita.py found: YES (/private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/numpy/_build_utils/tempita.py)
Configuring __config__.py using configuration
numpy/meson.build:208: WARNING: The variable(s) 'BLAS_INCLUDEDIR', 'BLAS_LIBDIR', 'BLAS_OPENBLAS_CONFIG', 'BLAS_PCFILEDIR', 'BLAS_TYPE_NAME', 'BLAS_VERSION', 'LAPACK_INCLUDEDIR', 'LAPACK_LIBDIR', 'LAPACK_OPENBLAS_CONFIG', 'LAPACK_PCFILEDIR', 'LAPACK_TYPE_NAME', 'LAPACK_VERSION' in the input file 'numpy/__config__.py.in' are not present in the given configuration data.
```

**Comment by rgommers** [comment]
@andyfaff FYI I make some minor tweaks to `cirrus_macosx_arm64.yml`. NumPy does not actually require a BLAS/LAPACK, and use of `CFLAGS` in CI jobs should be avoided, since it tends to mask problems that then will show up for users. With distutils there was no better (easy) way, but Meson has a `compiler.get_supported_arguments` method so that any needed compile flags can be added in a portable manner.

**Comment by rgommers** [comment]
A fun new error with `-Werror` in Python 3.11.3 that I cannot reproduce locally:
```
FAILED: numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so.p/src_umath_override.c.o 
cc -Inumpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so.p -Inumpy/core -I../numpy/core -Inumpy/core/include -I../numpy/core/include -I../numpy/core/src/common -I../numpy/core/src/multiarray -I../numpy/core/src/npymath -I../numpy/core/src/umath -I/usr/include/x86_64-linux-gnu/openblas-serial/ -I/opt/hostedtoolcache/Python/3.11.3/x64/include/python3.11 -fvisibility=hidden -fdiagnostics-color=always -Wall -Winvalid-pch -Werror -std=c99 -O2 -g -fno-strict-aliasing -DHAVE_CBLAS -fPIC -DNPY_INTERNAL_BUILD -DHAVE_NPY_CONFIG_H -DNPY_DISABLE_OPTIMIZATION -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -MD -MQ numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so.p/src_umath_override.c.o -MF numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so.p/src_umath_override.c.o.d -o numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so.p/src_umath_override.c.o -c ../numpy/core/src/umath/override.c
In file included from /opt/hostedtoolcache/Python/3.11.3/x64/include/python3.11/Python.h:44,
                 from ../numpy/core/include/numpy/npy_3kcompat.h:15,
                 from ../numpy/core/src/common/npy_pycompat.h:4,
                 from ../numpy/core/src/umath/override.c:4:
../numpy/core/src/umath/override.c: In function ‘PyUFunc_CheckOverride’:
/opt/hostedtoolcache/Python/3.11.3/x64/include/python3.11/object.h:538:9: error: ‘override_array_ufunc’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
  538 |         _Py_Dealloc(op);
      |         ^~~~~~~~~~~~~~~
../numpy/core/src/umath/override.c:319:19: note: ‘override_array_ufunc’ was declared here
  319 |         PyObject *override_array_ufunc;
      |                   ^~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

**Comment by seberg** [comment]
I think I saw it before at some point and I don't think the compiler is right, just guessing on the "this may happen" side of things.  Anyway, initializing to silence the warning is OK.

**Comment by rgommers** [comment]
> I don't think the compiler is right, just guessing on the "this may happen" side of things

Indeed, the code wasn't broken but the compiler cannot know. I think using an uninitialized variable that is then used only within an if-statement controlled by some other not-directly-the-same-pointer-condition should always be avoided.

**Comment by rgommers** [comment]
This is good to go now.

**Comment by mattip** [comment]
Thanks @rgommers

**Pull request merged**